uac-api/src/main/java/com/bazaarvoice/emodb/uac/api/AuthUserAccessControl.java

```diff
@@ -0,0 +1,141 @@
+package com.bazaarvoice.emodb.uac.api;
+
+import com.bazaarvoice.emodb.auth.proxy.Credential;
+import com.bazaarvoice.emodb.common.api.UnauthorizedException;
+
+import java.util.Iterator;
+
+/**
+ * Interface for user access control.  The methods in this interface are used for managing users, roles, and permissions.
+ * Note that use of these methods require special permissions from EmoDB.  If a caller invokes a method for which he has
+ * insufficient permissions the method will throw an {@link UnauthorizedException}.  Note that this exception takes
+ * precedence over any other exceptions or responses.  For example, assume a caller does not have permission to view
+ * any roles.  When that caller makes a call to {@link #getRole(String, EmoRoleKey)} he will always get an
+ * UnauthorizedException, regardless of whether the role he's getting actually exists or not.
+ *
+ * The first parameter to every call is the API key of the caller.  It is used for authenticating the caller and
+ * verifying his credentials.  Methods which manage API keys operate on the API key's ID, unless otherwise noted,
+ * and those parameters always appear after the caller's API key.
+ */
+public interface AuthUserAccessControl {
+
+    /**
+     * Gets all roles for which the user has read access.  This method will not throw an UnauthorizedException if
+     * a valid API key is used but does not have permission to view any roles.  In that case it will return
```

Comment:
It's because there is no overarching permission such as "user has permission to view roles".  Instead there are permissions to filter _which_ roles a user can read, such as `role|read|your_team|*`.  So functionally the service iterates over matching roles and filters out those for which the user doesn't have read permission.  As an optimization Emo could check whether any of the user's permissions give her permission to view any possible matching role and return immediately if not (e.g.; pre-check whether the user has any permission starting with`role|read`).  Even if this were implemented I'd think it would still return an empty list, otherwise the response would be very confusing.  Just because the caller doesn't have permission to view any existing roles today doesn't mean a new role won't be created tomorrow for which she does have read access.  So returning 403 today and a list with the one role she has read permission for tomorrow is inconsistent; only the data changed, not her permissions.